### DISCUSSION TITLE
How to debug generator issues?

### DISCUSSION THREAD

**Discussion by Marv51** [microsoft/CsWinRT#1787] [Q&A] [answered]
For one of my projects the generator throws this warning message: 
`CSC warning CS8785: Generator 'WinRTAotSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'.`

How do I find out more details? Can I get a stacktrace or attach a debugger somehow?

I captured a binlog, but it does not seem to contain any additional details:
[invalidIndex.zip](https://github.com/user-attachments/files/16963028/invalidIndex.zip)

I am setting:
```xml    
<CsWinRTAotWarningLevel>2</CsWinRTAotWarningLevel>
<CsWinRTEnableLogging>True</CsWinRTEnableLogging>
```
but I can not find any log.txt file.

I am sure I am doing something wrong, how can I find out what?

**Comment by manodasanW** [comment, marked as answer]
You might be able to attach a debugger to one of the Rosyln processes.  But I will look at adding some logs or diagnostics when exceptions get thrown in the source generator and then you can try that out.

**Reply by Marv51** [reply]
Update: I thought this message disappeared, but it is back now. Not sure if it was ever gone, or if I just couldn't find it. 
Sorry for the noise. Some logs or diagnostics would still be really appreciated. But I love all the hard work you are putting into this project. 👍🙏



Side note: There is a typo in the error message in 
https://github.com/microsoft/CsWinRT/blob/3478372c5511918c723a08749b67c1ffbd12db52/src/WinRT.Runtime/Projections/ICustomPropertyProvider.net5.cs#L411
I believe it should be:
`"ICustomProperty support used by XAML binding for '{target.GetType()}' requires the type to **be** marked with the 'WinRT.GeneratedBindableCustomPropertyAttribute'."`
(I don't want to create a noisy PR or Issue, for such a nit)

**Reply by manodasanW** [reply]
I am working on figuring out the appropriate places for us to put diagnostics and make it actionable.  But in the meantime, there might be something you can try that would give us a hint on the scenario where this is happening.  In Visual Studio, can you go to `Tools -> Options -> Projects and Solutions -> Build and Run -> MSBuild project build output verbosity` and set it to `Diagnostic`.  After that, can you repro with a binlog and share that.  You might also just be able to just share the stack you should see in the build output itself if you search for `'WinRTAotSourceGenerator' failed to generate source`.  After you do that, you can undo the setting to what you had previously.

And thanks for noting the typo, will get it fixed.

**Reply by Marv51** [reply]
This is useful, thanks for the instructions. I'll share this stack here I'll look into it further, trying to get a better repro or getting binlogs.

```
1>    CSC : warning CS8785: Generator 'WinRTAotSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'IndexOutOfRangeException' with message 'Index was outside the bounds of the array.'.
1>    System.IndexOutOfRangeException: Index was outside the bounds of the array.
1>       at Generator.WinRTAotSourceGenerator.GetVtableAttributesToAddOnLookupTable(GeneratorSyntaxContext context, TypeMapper typeMapper, Func`3 isWinRTType, Func`2 isWinRTClass)
1>       at Generator.WinRTAotSourceGenerator.GetVtableAttributesToAddOnLookupTable(GeneratorSyntaxContext context, TypeMapper typeMapper, Boolean isCsWinRTComponent)
1>       at Generator.WinRTAotSourceGenerator.<>c.<Initialize>b__0_16(ValueTuple`2 value, CancellationToken _)
1>       at Microsoft.CodeAnalysis.TransformNode`2.<>c__DisplayClass6_0.<.ctor>b__0(TInput i, CancellationToken token)
1>       at Microsoft.CodeAnalysis.TransformNode`2.UpdateStateTable(Builder builder, NodeStateTable`1 previousTable, CancellationToken cancellationToken)
1>    -----
```

**Reply by dongle-the-gadget** [reply]
Looks like the same error as #1828.

**Comment by Marv51** [comment]
Latest update fixed my issue. Thank you @manodasanW and team!
